File: nextGreaterElements.py
# Time Complexity : O(n)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        numLen = len(nums)
        nums = nums + nums

        n = len(nums)
        res = [0] * n
        stack = []
        for i in range(n-1, -1, -1):
            while stack and nums[i] >= stack[-1]:
                stack.pop()
            if stack:
                res[i] = stack[-1]
            else:
                res[i] = -1
            stack.append(nums[i])
        return res[:n//2]

File: nextGreaterElements.cpp
// Time Complexity : O(n) => O(n+n) cauae of the for looop
// Space Complexity : O(n) all elements can go into the satck
// Did this code successfully run on Leetcode : Yes


/*
Monotonic stacks
- Everytime you see a number bigger than the stack.top(), update the result 
- Go over the array twice to capture the greatest circularly
- Don't add the element one the second iteration of the array since the result has already been ipdated 
*/

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        if(nums.size() == 0) return nums;
        int n = nums.size();
        // initalise with -1
        vector<int> result (n, -1);

        stack<int> st;
        // iterating over the array twice
        for(int i = 0; i < 2 * n; i++){
            // If you see a new number bigger than the current top, keep updatating the result
            while(!st.empty() && nums[i%n] > nums[st.top()]) {
                result[st.top()] = nums[i%n];
                st.pop();
            }

            // if its the first iteration. push the number to the stack
            if(i < n)
                st.push(i);
        }
        return result;
    }
};